### DISCUSSION TITLE
How to read the `javax.persistence.Column.length()` value during test data generation?

### DISCUSSION THREAD

**Discussion by nedis-tech** [instancio/instancio#467] [Q&A] [answered]
Hello,

I would like to generate JPA entities with string test data that are valid according to "`@Column(length = x)`" restrictions.
I.e. something like this:

```
return Instancio.of(entityClass)
    // Try to parse @Column.length() value for test string data:
    .generate(
        fields(field -> field.getType().equals(String.class) && field.isAnnotationPresent(Column.class)),
        gen -> {
            Field field = //TODO getCurrentField instance
            return gen.string().maxLength(field.getAnnotation(Column.class).length());
        })
    .create();
```
Is it possible to set such string generator?

**Comment by armandino** [comment, marked as answer]
Hi @nedis-ua, unfortunately there's no way to pass an annotation attribute value from a selector to the generator.

There is experimental support for bean validation annotations (including some Hibernate-specific ones):
https://www.instancio.org/user-guide/#bean-validation

Can I ask what is the goal you're trying to accomplish? Can the generated string be shorter than the `length` attribute for your test case, or do you need the exact `length`?

**Reply by nedis-tech** [reply]
Hi, @armandino , Thank you for answer.

My goal is to create test instances of JPA entities that can be stored to database without SQL error.

For example:
```
class Entity {
    @Column(length = 1)
    String flag;

    @Column(length = 100)
    String value;
}
```

Therefore the max length for first field is `1` and the max length for the second one is `100`. 

If Instancio API would have an ability to read annotations from fields, I could configure the string generator according to length attribute values.

**Reply by armandino** [reply]
@nedis-ua, thank you for the clarification. I think it would be a useful feature to handle this generically, as you described in your original post. I will need to experiment a little with the API, but this might take me some time. I created an issue (#474) to investigate it further.

In the meantime, if you have entities with many short columns, one option would be to set the `Keys.STRING_MAX_LENGTH` to a smaller value (the default is 10) using  the `Settings` class. This might reduce the number of fields you have to set manually via `generate()`.

**Reply by nedis-tech** [reply]
@armandino , 

As for `Keys.STRING_MAX_LENGTH` I know about this feature, but I am going to wait until the https://github.com/instancio/instancio/issues/474 issue will be resolved

**Reply by armandino** [reply]
@nedis-ua I added a new feature to support this functionality. It's documented here https://www.instancio.org/user-guide/#jpa and is available from version `3.3.0`.
